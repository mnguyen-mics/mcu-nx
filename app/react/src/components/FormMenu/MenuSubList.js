
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import McsIcons from '../McsIcons';

class MenuList extends Component {

  constructor(props) {
    super(props);
    this.state = {
      open: false,
    };
  }

  openMenuItem = () => {
    return this.setState({
      open: !this.state.open,
    });
  }

  render() {
    const {
      title,
      subtitles,
      submenu,
    } = this.props;
    return (
      <div className="menu-sublist">
        <button className={this.state.open ? 'menu-item opened' : 'menu-item'} onClick={this.openMenuItem}>
          <div className={subtitles ? 'content' : 'content alone'}>
            <div className="title">{title}</div>
            {subtitles ? <div className="subtitles">{ subtitles.map((subtitle, index) => {
              return index !== subtitles.length - 1 ? `${subtitle}, ` : subtitle;
            }) }</div> : null}
          </div>
          <div className="selector">
            <McsIcons type={this.state.open ? 'minus' : 'plus'} />
          </div>
        </button>
        <div className={`lines ${this.state.open ? 'opened' : 'closed'}`}>
          {submenu && submenu.map(sub => {
            return (
              <button key={sub.title} className="menu-item small" onClick={sub.select}>
                <div className="content alone small">
                  <div className="subtitles">{sub.title}</div>
                </div>
                <div className="selector"><McsIcons type={'chevron-right'} /></div>
              </button>
            );
          })}
        </div>
      </div>
    );
  }
}

MenuList.defaultProps = {
  subtitles: null
};

MenuList.propTypes = {
  title: PropTypes.string.isRequired,
  subtitles: PropTypes.arrayOf(PropTypes.string.isRequired),
  submenu: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string.isRequired,
    select: PropTypes.func.isRequired,
  }).isRequired).isRequired,
};

export default MenuList;
