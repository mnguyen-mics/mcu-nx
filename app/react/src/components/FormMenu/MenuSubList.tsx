
import React, { Component } from 'react';
import McsIcons from '../McsIcons';

interface Submenu {
  title: string;
  select: React.FormEventHandler<any>;
}

export interface MenuListProps {
  title: string;
  subtitles: string[];
  submenu: Submenu[];
}

export interface MenuListState {
  open: boolean;
}

class MenuList extends Component<MenuListProps, MenuListState> {

  constructor(props: MenuListProps) {
    super(props);
    this.state = {
      open: false,
    };
  }

  openMenuItem = () => {
    return this.setState({
      open: !this.state.open,
    });
  }

  render() {
    const {
      title,
      subtitles,
      submenu,
    } = this.props;
    return (
      <div className="menu-sublist">
        <button className={this.state.open ? 'menu-item opened' : 'menu-item'} onClick={this.openMenuItem}>
          <div className={subtitles ? 'content' : 'content alone'}>
            <div className="title">{title}</div>
            {subtitles ? <div className="subtitles">{ subtitles.map((subtitle, index) => {
              return index !== subtitles.length - 1 ? `${subtitle}, ` : subtitle;
            }) }</div> : null}
          </div>
          <div className="selector">
            <McsIcons type={this.state.open ? 'minus' : 'plus'} />
          </div>
        </button>
        <div className={`lines ${this.state.open ? 'opened' : 'closed'}`}>
          {submenu && submenu.map(sub => {
            return (
              <button key={sub.title} className="menu-item small" onClick={sub.select}>
                <div className="content alone small">
                  <div className="subtitles">{sub.title}</div>
                </div>
                <div className="selector"><McsIcons type={'chevron-right'} /></div>
              </button>
            );
          })}
        </div>
      </div>
    );
  }
}

export default MenuList;
