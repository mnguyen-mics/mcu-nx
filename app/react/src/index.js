import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';

import configureStore from './store';
import { Navigator } from './containers/Navigator';

const store = configureStore();

const micsProvider = (
  <Provider store={store}>
    <Navigator />
  </Provider>
);

render(
  micsProvider,
  document.getElementById('mcs-react-app') // eslint-disable-line no-undef
);
