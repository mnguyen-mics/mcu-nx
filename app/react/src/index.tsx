/* eslint-disable */
import React from 'react';
import { render } from 'react-dom';

import App from './App';

declare global {
  interface Window {
    require: any;
    mainRev: any;
  }
}

render(<App />, document.getElementById('mcs-react-app'), () => {
  window.require([window.mainRev], () => {
    //
  });
});
