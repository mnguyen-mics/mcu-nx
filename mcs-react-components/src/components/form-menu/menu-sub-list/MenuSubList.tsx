import * as React from 'react';
import * as cuid from 'cuid';
import { Spin } from 'antd';
import { FormattedMessage } from 'react-intl';
import McsIcon from '../../mcs-icon';

export interface SubMenu {
  title: string;
  select: React.FormEventHandler<any>;
}

export interface MenuSubListProps {
  className?: string;
  title: string;
  subtitles: string[];
  submenu: SubMenu[] | (() => Promise<SubMenu[]>);
}

export interface MenuSubListState {
  open: boolean;
  fetchedMenu: SubMenu[];
  fetching: boolean;
  fetched: boolean;
}

type Props = MenuSubListProps;

class MenuSubList extends React.Component<Props, MenuSubListState> {
  constructor(props: Props) {
    super(props);
    this.state = {
      open: false,
      fetching: false,
      fetched: false,
      fetchedMenu: [],
    };
  }

  fetchSubmenu = () => {
    const { submenu } = this.props;
    if (typeof submenu === 'function') {
      this.setState({ fetching: true });
      submenu()
        .then(menu => {
          this.setState({
            fetching: false,
            fetched: true,
            fetchedMenu: menu,
          });
        })
        .catch(() => {
          this.setState({ fetching: false });
        });
    }
  };

  openMenuItem = () => {
    const { fetched } = this.state;
    if (!fetched) this.fetchSubmenu();
    this.setState({
      open: !this.state.open,
    });
  };

  render() {
    const { title, subtitles, submenu, className } = this.props;

    const { fetching, fetchedMenu } = this.state;

    let displayMenu: SubMenu[] = fetchedMenu;
    if (Array.isArray(submenu)) {
      displayMenu = submenu;
    }

    return (
      <div className={`mcs-menu-sub-list ${className ? className : ''}`}>
        <button
          className={this.state.open ? 'menu-item opened' : 'menu-item'}
          onClick={this.openMenuItem}
        >
          <div className={subtitles ? 'content' : 'content alone'}>
            <div className='title'>{title}</div>
            {subtitles ? (
              <div className='subtitles'>
                {subtitles.map((subtitle, index) => {
                  return index !== subtitles.length - 1 ? `${subtitle}, ` : subtitle;
                })}
              </div>
            ) : null}
          </div>
          <div className='selector'>
            <McsIcon type={this.state.open ? 'minus' : 'plus'} />
          </div>
        </button>
        {this.state.open && (
          <div>
            {fetching && (
              <div className='menu-item small'>
                <div className='content alone small text-center'>
                  <Spin />
                </div>
              </div>
            )}
            {!fetching && !displayMenu.length && (
              <div className='menu-item small'>
                <div className='content alone small text-center'>
                  <FormattedMessage
                    id='components.formMenu.menuSubList.empty'
                    defaultMessage='Empty'
                  />
                </div>
              </div>
            )}
            {!fetching &&
              displayMenu.map(sub => {
                return (
                  <button key={cuid()} className='menu-item small' onClick={sub.select}>
                    <div className='content alone small'>
                      <div className='subtitles'>{sub.title}</div>
                    </div>
                    <div className='selector'>
                      <McsIcon type={'chevron-right'} />
                    </div>
                  </button>
                );
              })}
          </div>
        )}
      </div>
    );
  }
}

export default MenuSubList;
