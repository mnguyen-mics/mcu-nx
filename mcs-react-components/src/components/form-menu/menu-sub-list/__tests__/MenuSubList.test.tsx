import 'jest';
import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import MenuSubList, { MenuSubListProps } from '../MenuSubList';

it('renders the MenuSubList', () => {
  const props: MenuSubListProps = {
    title: 'Title',
    subtitles: ['subtitle_1', 'subtitle_2'],
    submenu: [{
      title: 'submenu_1',
      select: () => { /* tslint:disable */ console.log('click submenu_1') /* tslint:enable */ },
    }],
  };
  const component = TestRenderer.create(
    <MenuSubList {...props}/>
  );
  const res = component.toJSON();
  expect(res).toMatchSnapshot();
});

it('renders the MenuSubList with submenu', () => {
  const props: MenuSubListProps = {
    title: 'Title',
    subtitles: ['subtitle_1', 'subtitle_2'],
    submenu: [{
      title: 'submenu_1',
      select: () => { /* tslint:disable */ console.log('click submenu_1') /* tslint:enable */ },
    }],
  };
  const component = TestRenderer.create(
    <MenuSubList {...props}/>
  );
  component.root.findByType('button').props.onClick();
  const res = component.toJSON();
  expect(res).toMatchSnapshot();
});
