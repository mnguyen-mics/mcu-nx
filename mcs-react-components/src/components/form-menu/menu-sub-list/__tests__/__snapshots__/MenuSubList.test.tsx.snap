// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the MenuSubList 1`] = `
<div
  className="mcs-menu-sub-list"
>
  <button
    className="menu-item"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div
        className="title"
      >
        Title
      </div>
      <div
        className="subtitles"
      >
        subtitle_1, 
        subtitle_2
      </div>
    </div>
    <div
      className="selector"
    >
      <span
        className="icon "
      >
        <i
          className="mcs-plus"
        />
      </span>
    </div>
  </button>
</div>
`;

exports[`renders the MenuSubList with submenu 1`] = `
<div
  className="mcs-menu-sub-list"
>
  <button
    className="menu-item opened"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div
        className="title"
      >
        Title
      </div>
      <div
        className="subtitles"
      >
        subtitle_1, 
        subtitle_2
      </div>
    </div>
    <div
      className="selector"
    >
      <span
        className="icon "
      >
        <i
          className="mcs-minus"
        />
      </span>
    </div>
  </button>
  <div>
    <button
      className="menu-item small"
      onClick={[Function]}
    >
      <div
        className="content alone small"
      >
        <div
          className="subtitles"
        >
          submenu_1
        </div>
      </div>
      <div
        className="selector"
      >
        <span
          className="icon "
        >
          <i
            className="mcs-chevron-right"
          />
        </span>
      </div>
    </button>
  </div>
</div>
`;
